lib/vm/src/mmap.rs

```diff
@@ -66,15 +99,37 @@ impl Mmap {
             return Ok(Self::new());
         }
 
+        // Open a temporary file (which is used for swapping)
+        let fd = unsafe {
+            let file = libc::tmpfile();
+            if file.is_null() {
+                return Err(format!(
+                    "failed to create temporary file - {}",
+                    io::Error::last_os_error()
+                ));
+            }
+            FdGuard(libc::fileno(file))
+        };
+
+        // First we initialize it with zeros
+        unsafe {
+            if libc::ftruncate(fd.0, mapping_size as libc::off_t) < 0 {
+                return Err("could not truncate tmpfile".to_string());
+            }
+        }
+
+        // Compute the flags
+        let flags = libc::MAP_FILE | libc::MAP_SHARED;
+
         Ok(if accessible_size == mapping_size {
             // Allocate a single read-write region at once.
             let ptr = unsafe {
                 libc::mmap(
                     ptr::null_mut(),
                     mapping_size,
                     libc::PROT_READ | libc::PROT_WRITE,
-                    libc::MAP_PRIVATE | libc::MAP_ANON,
-                    -1,
+                    flags,
+                    fd.0,
```

Comment:
Doesn't the systematic file mapping will slowdown things (compared to previous Private Anonymous map)?

And I'm still unsure of the purpose of this.

Reply:
Even for a suspend on disk, I think the performance penalty is not worth it. The suspend on disk is a "rare" operation compare to read/write memory.

Reply:
Suspend to disk will be very common on Deploy servers, because many instances will be idle for a long time, so we want to evict them from memory.

You can actually configure a mapping so it will only sync to disk when there is memory pressure.
I'm not sure yet how well this will work in practice but that's the plan.

Either way it should not be the default.

We either need a way to configure the mmap behaviour, or we just duplicate the OwnedMemory and SharedMemory in deploy.

Reply:
Probably better to just duplicate for now, it's not a whole lot of code.

Reply:
Actually, that might require us adding a way to downcast memory to a concrete type (by adding something like `LinearMemory::as_any(&self) -> Option<&dyn Any>`.

Reply:
as discussed on notion - no problem with having multiple implementations of the memory but for Deploy we need high density (and so a backing file)

Reply:
FD MMap implementation was removed from `vm`. 

A separate memory implementation was added in `wasi`.